### PR TITLE
Cygwin support

### PR DESCRIPTION
**Pull request by benrubson**
Hi,

This PR adds support to compile and run EncFS on Cygwin.
The FUSE layer is handled by [WinFsp](https://github.com/billziss-gh/winfsp).
Port history can be found [here](https://github.com/billziss-gh/winfsp/issues/152).

All tests pass, but :
- in normal mode, symlink tests ;
- in reverse mode, absolute symlink tests.

The first point is due to how Cygwin handles symlinks. It uses small files with content beginning with a specific pattern, followed by the link target. Encryption breaks this. Perhaps we could think later on about a passthrough mode for these files.
The second point may be on WinFsp side, https://github.com/billziss-gh/winfsp/issues/153 will tell us more.
_In other words, we can read existing relative symlinks, but can't create symlinks._

See [the wiki](https://github.com/vgough/encfs/wiki) for installation instructions.

WinFsp may give nice performance improvements possibilities, [especially thanks to `READDIR_PLUS`](https://github.com/billziss-gh/winfsp/issues/44).
Some work may however be needed to use the WIN `readdir` API.

Thank you 👍 

Ben

_This PR depends on #494._

sc

### CODE REVIEW

File: encfs/DirNode.cpp [hunk 1]
```diff
@@ -739,8 +739,15 @@ int DirNode::unlink(const char *plaintextName) {
...
+  size_t maxopen = 0;
+#ifdef __CYGWIN__
+  // When deleting a file, Windows first opens it, deletes it, and closes it
+  // We then have inevitably the file opened one time 
+  maxopen = 1;
+#endif
+
   int res = 0;
-  if ((ctx != nullptr) && ctx->lookupNode(plaintextName)) {
+  if ((ctx != nullptr) && (ctx->numberOfNode(plaintextName) > maxopen)) {
```
Comment (rfjakob) [inline encfs/DirNode.cpp]: Windows does not allow deleting open files, so we can just skip the check.
In other words:
```
if ( !windows && (ctx != nullptr) && (ctx->numberOfNode(plaintextName) > maxopen)) {
```

File: encfs/FileUtils.cpp [hunk 2]
```diff
@@ -1739,6 +1739,9 @@ void unmountFS(const char *mountPoint) {
   // fuse_unmount does not work on Mac OS, see #428
   unmount(mountPoint, MNT_FORCE);
 #endif
+#ifdef __CYGWIN__
+  system(string("/usr/bin/pkill -f '(^|/)encfs .*/.* ").append(mountPoint).append("?( |$)'").c_str());
```
Comment (rfjakob) [inline encfs/FileUtils.cpp]: Shell injection vulnerabilty!

Comment (benrubson) [reply encfs/FileUtils.cpp]: Thank you for your review 👍 
I think a regex match of `mountPoint` would do the trick ?
Perhaps you thought about something else ?

Comment (rfjakob) [reply encfs/FileUtils.cpp]: The best way is to avoid the shell by using something like execl

Comment (benrubson) [reply encfs/FileUtils.cpp]: Done 👍

File: encfs/DirNode.cpp (hunk 1 above)
Comment (rfjakob) [reply encfs/DirNode.cpp]: I asked myself the same thing yesterday, but this code is even older than the git history, and I am not completely sure. It may be related to how encfs shares backing file descriptors between multiple open FUSE file descriptors. I believe they are stored in a table indexed by path.

File: encfs/FileUtils.cpp [hunk 3]
```diff
@@ -1739,6 +1739,12 @@ void unmountFS(const char *mountPoint) {
   // fuse_unmount does not work on Mac OS, see #428
   unmount(mountPoint, MNT_FORCE);
 #endif
+#ifdef __CYGWIN__
+  if(fork() == 0)
```
Comment (benrubson) [inline encfs/FileUtils.cpp]: I think parent should way its child, because parent exits before `pkill` completes, and it makes some tests to fail.
I will then correct this.

**Comment by benrubson** [comment]
Thank you for your review & approval @rfjakob 👍

File: encfs/FileUtils.cpp (hunk 3 above)
Comment (benrubson) [reply encfs/FileUtils.cpp]: Done
